Add size_hints to cache key

PaulZhang12 opened the pull request:
Differential Revision: D78089705

Previously to support overriding autotune configs for post fusion kernels in Inductor with a lookup table, we only keyed on the source code. However, the same source code could have multiple optimal configs, due to the input sizes. With this, we have many collisions in our lookup table, leading to subpar configs. A way around this is to add the size_hints to the lookup key as well


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Skylion007 commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -1156,6 +1155,8 @@ def benchmark_one_config(config):
                 best_config
             ).make_launcher()
+
+        log.error(f"Function hash {generate_lookup_hash_from_source_code(str(self.size_hints), self.fn.src)} has best config {best_config}")
```
```suggestion
        log.error("Function hash %s has best config %s", fn_hash, best_config)
```
Is a lazily evaluated string and way more efficient, particularly if the user does not want to actually log the data.

PaulZhang12 committed:
[Inductor] Add size_hints to autotune override lookup key (#158026)

Summary:

Previously to support overriding autotune configs for post fusion kernels in Inductor with a lookup table, we only keyed on the source code. However, the same source code could have multiple optimal configs, due to the input sizes.

Below is a kernel from production that shows this phenomenon.

```
triton_heuristics.pointwise(
    size_hints={'x': 32768},
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*i64', 'out_ptr0': '*i64', 'ks0': 'i64', 'xnumel': 'i32', 'XBLOCK': 'constexpr'}, 'device': DeviceProperties(type='cuda', index=1, multi_processor_count=132, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, warp_size=32), 'constants': {}, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid1D', 'autotune_hints': set(), 'kernel_name': 'triton_poi_fused__to_copy_cat_3', 'mutated_arg_names': [], 'optimize_mem': False, 'no_x_dim': False, 'num_load': 2, 'num_reduction': 0, 'backend_hash': '07B1197BCEA79355F58E697BC7E8ADF4BFB6253A31574E91603DE742DF281D67', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'is_fbcode': True, 'coordinate_descent_tuning': True, 'coordinate_descent_search_radius': 1, 'coordinate_descent_check_all_directions': False},
    min_elem_per_thread=0
)
triton.jit
def triton_poi_fused__to_copy_cat_3(in_ptr0, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = x0
    tmp1 = tl.full([1], 0, tl.int64)
    tmp2 = tmp0 >= tmp1
    tmp3 = (-1) + ks0
    tmp4 = tmp0 < tmp3
    tmp5 = tl.load(in_ptr0 + (x0), tmp4 & xmask, eviction_policy='evict_last', other=0.0)
    tmp6 = tl.load(in_ptr0 + (1 + (x0)), tmp4 & xmask, eviction_policy='evict_last', other=0.0)
    tmp7 = tmp5 - tmp6
    tmp8 = tmp7.to(tl.float32)
    tmp9 = tl.full(tmp8.shape, 0.0, tmp8.dtype)
    tmp10 = tl.where(tmp4, tmp8, tmp9)
    tmp11 = tmp0 >= tmp3
    tmp12 = ks0
    tmp13 = tmp0 < tmp12
    tmp14 = 0.0
    tmp15 = tl.full(tmp14.shape, 0.0, tmp14.dtype)
    tmp16 = tl.where(tmp11, tmp14, tmp15)
    tmp17 = tl.where(tmp4, tmp10, tmp16)
    tmp18 = tmp17.to(tl.int64)
    tl.store(out_ptr0 + (x0), tmp18, xmask)
```

This kernel produces the hash `7629762dd891afea800eb5d26d88be60d7e3d56e8897f52a666e9ba76da02f99`, as a hash of the signature and the source code of the kernel. However, the size hints on this kernel can change significantly on recompilations, leading to different optimal configs. With this, we have many collisions in our lookup table, leading to subpar configs. A way around this is to add the size_hints to the lookup key as well, which decreases the amount of collisions for this kernel on the OmniFM model from 16 -> 5

Test Plan:
`buck2 run mode/opt caffe2/test/inductor:async_compile`

Rollback Plan:

Differential Revision: D78089705
